File: plugins/mission-control/flow-server/src/domain/annotation.rs
//! Pure annotation formatter (roadmap #4 comment loop) — no IO. Turns a human
//! comment on an item into a deterministic, byte-stable markdown block that the
//! store appends to the item's plan document. Keeping the formatter pure means
//! the only IO (resolving the plan path and appending) lives in [`crate::store`],
//! and the block layout is pinned by tests independently of the file write.

use super::ids::ItemId;

/// Format a human comment as a markdown annotation block for an item's plan.
///
/// The block is self-contained and append-safe: it opens with a blank line so it
/// never runs into preceding content, carries a heading naming the item, and
/// renders the comment verbatim. The output is deterministic for a given
/// `(id, comment)` pair so it can be diffed and tested.
pub fn format_annotation(id: &ItemId, comment: &str) -> String {
    let body = comment.trim_end();
    format!(
        "\n<!-- annotation: {id} -->\n### Annotation on `{id}`\n\n{body}\n",
        id = id.as_str(),
        body = body,
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    fn id(s: &str) -> ItemId {
        ItemId::new(s).unwrap()
    }

    #[test]
    fn formats_a_simple_comment() {
        let out = format_annotation(&id("flow-server"), "ship it");
        assert_eq!(
            out,
            "\n<!-- annotation: flow-server -->\n### Annotation on `flow-server`\n\nship it\n"
        );
    }

    #[test]
    fn trims_only_trailing_whitespace() {
        // Leading whitespace is preserved (it may be meaningful markdown); trailing
        // whitespace/newlines are trimmed so the block ends with exactly one '\n'.
        let out = format_annotation(&id("a"), "  indented body  \n\n");
        assert_eq!(
            out,
            "\n<!-- annotation: a -->\n### Annotation on `a`\n\n  indented body\n"
        );
    }

    #[test]
    fn empty_comment_yields_empty_body_line() {
        let out = format_annotation(&id("a"), "");
        assert_eq!(out, "\n<!-- annotation: a -->\n### Annotation on `a`\n\n\n");
    }

    #[test]
    fn multiline_comment_is_rendered_verbatim() {
        let out = format_annotation(&id("a"), "line one\nline two");
        assert!(out.contains("line one\nline two\n"));
        assert!(out.ends_with("line one\nline two\n"));
    }

    #[test]
    fn is_byte_stable_across_calls() {
        assert_eq!(
            format_annotation(&id("a"), "same"),
            format_annotation(&id("a"), "same")
        );
    }
}

File: plugins/flow/flow-mcp/lib/flow_mcp/annotation.rb
# frozen_string_literal: true

module FlowMcp
  # Pure annotation formatter (EARS-FLOW-061) — no IO. Turns a human comment into
  # the deterministic, append-safe markdown block the store appends to an item's
  # plan document. Byte-identical to the retired Rust reference.
  module Annotation
    module_function

    def format(id, comment)
      body = comment.rstrip
      "\n<!-- annotation: #{id} -->\n### Annotation on `#{id}`\n\n#{body}\n"
    end

    # Derive the plan-doc filename for an item title (EARS-FLOW-060): uppercase,
    # each run of non-alphanumerics collapsed to a single `_`, surrounding `_`
    # trimmed, suffixed `_PLAN.md`. e.g. "Flow server" -> "FLOW_SERVER_PLAN.md".
    def plan_doc_filename(title)
      name = +""
      last_us = false
      title.each_char do |ch|
        if ch.match?(/[A-Za-z0-9]/)
          name << ch.upcase
          last_us = false
        elsif !last_us
          name << "_"
          last_us = true
        end
      end
      "#{name.gsub(/\A_+|_+\z/, '')}_PLAN.md"
    end
  end
end
